File: server/src/services/pondSiteSelection.js
/**
 * Pond Site Selection Service
 * 
 * Identifies suitable LAND locations for village pond construction based on
 * multi-factor terrain suitability analysis.
 * 
 * Key design principle: The pond should be on land ADJACENT to the drainage
 * system, NOT directly on the strongest drainage channel/stream/river.
 * 
 * Selection criteria:
 * - Local depression (terrain dips below surroundings)
 * - Low relative elevation (lower portions of terrain)
 * - Low/moderate slope (flat enough for pond construction)
 * - Meaningful upstream catchment (enough water will reach the pond)
 * - Terrain convergence (surrounded by higher ground)
 * - Offset from the main drainage channel (NOT on the river/stream itself)
 * 
 * The scoring weights are configurable constants so the system can be tuned
 * for different contour maps without hard-coding sample-specific values.
 * 
 * Returns ranked candidates with composite scores and detailed breakdowns.
 */

const { gridToCoords, haversineDistance } = require('../utils/geometry');
const { DR, DC } = require('./terrainAnalysis');

// ─────────────────────────────────────────────────────────────────────────────
// CONFIGURABLE SCORING WEIGHTS
// Adjust these to change the relative importance of each suitability factor.
// All weights should sum to 1.0.
// ─────────────────────────────────────────────────────────────────────────────
const WEIGHTS = {
  elevation:     0.10,  // Prefer lower relative elevation
  slope:         0.15,  // Prefer flat/low-slope areas
  depression:    0.20,  // Prefer local depressions (terrain bowls)
  convergence:   0.10,  // Prefer areas surrounded by higher terrain
  catchment:     0.20,  // Prefer areas with meaningful upstream catchment
  channelOffset: 0.25,  // Penalize cells on or very near drainage channels
};

// Percentile threshold for identifying "drainage channel" cells.
// Cells with flow accumulation in the top CHANNEL_PERCENTILE fraction
// are classified as channel/stream cells.
const CHANNEL_PERCENTILE = 0.02; // top 2% of accumulation values

// Minimum flow accumulation fraction required for a candidate.
// Cells below this fraction of the max accumulation are skipped.
const MIN_ACCUMULATION_FRACTION = 0.005; // 0.5% of max

// Maximum relative elevation for candidates (0 = lowest, 1 = highest).
const MAX_RELATIVE_ELEVATION = 0.65;

// Neighborhood radius for depression detection (in grid cells).
// A 5×5 window uses radius=2, 7×7 uses radius=3, etc.
const DEPRESSION_RADIUS = 3;

// Minimum channel offset distance in grid cells for ideal score.
// Cells this far or farther from the drainage channel get full offset score.
const IDEAL_CHANNEL_OFFSET = 5;

// Maximum number of candidates to return (including selected).
const MAX_CANDIDATES = 5;

// ─────────────────────────────────────────────────────────────────────────────

/**
 * Select candidate pond locations from terrain model.
 * @param {Object} terrainModel - Output from buildTerrainModel
 * @param {Object} metadata - Contour metadata
 * @returns {Object} Selected pond site and candidates with score breakdowns
 */
function selectPondSite(terrainModel, metadata) {
  const {
    elevationGrid, flowAccumulation, flowDirection,
    nRows, nCols, bounds, cellSizeLng, cellSizeLat, cellSizeMeters
  } = terrainModel;

  const elevMin = metadata.minElevation;
  const elevMax = metadata.maxElevation;
  const elevRange = elevMax - elevMin || 1;

  // ── Step 1: Compute global flow accumulation statistics ──────────────
  const accumValues = [];
  let maxAccum = 0;
  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      const a = flowAccumulation[row][col];
      accumValues.push(a);
      if (a > maxAccum) maxAccum = a;
    }
  }

  // Sort to find the channel threshold at the given percentile
  accumValues.sort((a, b) => a - b);
  const channelThresholdIndex = Math.floor(accumValues.length * (1 - CHANNEL_PERCENTILE));
  const channelThreshold = accumValues[Math.min(channelThresholdIndex, accumValues.length - 1)];

  console.log(`  Flow accumulation: max=${maxAccum.toFixed(0)}, channel threshold (top ${CHANNEL_PERCENTILE * 100}%)=${channelThreshold.toFixed(0)}`);

  // ── Step 2: Build drainage channel mask ──────────────────────────────
  // A cell is classified as "channel" if its accumulation >= channelThreshold
  const isChannel = Array.from({ length: nRows }, () => new Uint8Array(nCols));
  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      if (flowAccumulation[row][col] >= channelThreshold) {
        isChannel[row][col] = 1;
      }
    }
  }

  // ── Step 3: Compute distance-to-nearest-channel for every cell ──────
  // BFS from all channel cells simultaneously
  const channelDist = Array.from({ length: nRows }, () =>
    new Float32Array(nCols).fill(Infinity)
  );
  const bfsQueue = [];
  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      if (isChannel[row][col]) {
        channelDist[row][col] = 0;
        bfsQueue.push(row * nCols + col);
      }
    }
  }
  let head = 0;
  while (head < bfsQueue.length) {
    const idx = bfsQueue[head++];
    const r = Math.floor(idx / nCols);
    const c = idx % nCols;
    const currentDist = channelDist[r][c];
    for (let d = 0; d < 8; d++) {
      const nr = r + DR[d];
      const nc = c + DC[d];
      if (nr < 0 || nr >= nRows || nc < 0 || nc >= nCols) continue;
      const stepDist = (d % 2 === 0) ? 1.0 : 1.414; // diagonal is √2
      const newDist = currentDist + stepDist;
      if (newDist < channelDist[nr][nc]) {
        channelDist[nr][nc] = newDist;
        bfsQueue.push(nr * nCols + nc);
      }
    }
  }

  // ── Step 4: Compute slope grid ──────────────────────────────────────
  // Slope = max elevation drop per unit distance across 8-neighbors
  const slopeGrid = Array.from({ length: nRows }, () => new Float64Array(nCols));
  let maxSlope = 0;
  for (let row = 0; row < nRows; row++) {
    for (let col = 0; col < nCols; col++) {
      const elev = elevationGrid[row][col];
      let maxGradient = 0;
      for (let d = 0; d < 8; d++) {
        const nr = row + DR[d];
        const nc = col + DC[d];
        if (nr < 0 || nr >= nRows || nc < 0 || nc >= nCols) continue;
        const dist = (d % 2 === 0) ? 1.0 : 1.414;
        const gradient = Math.abs(elev - elevationGrid[nr][nc]) / dist;
        if (gradient > maxGradient) maxGradient = gradient;
      }
      slopeGrid[row][col] = maxGradient;
      if (maxGradient > maxSlope) maxSlope = maxGradient;
    }
  }

  // ── Step 5: Score every candidate cell ──────────────────────────────
  const margin = Math.max(DEPRESSION_RADIUS + 1, Math.floor(Math.min(nRows, nCols) * 0.05));
  const minAccum = maxAccum * MIN_ACCUMULATION_FRACTION;
  const logMaxAccum = Math.log(maxAccum + 1);
  const cellScores = [];

  for (let row = margin; row < nRows - margin; row++) {
    for (let col = margin; col < nCols - margin; col++) {
      const elev = elevationGrid[row][col];
      const accum = flowAccumulation[row][col];
      const slope = slopeGrid[row][col];
      const distToChannel = channelDist[row][col];

      // ─ Filter: skip cells at high elevations
      const relativeElev = (elev - elevMin) / elevRange;
      if (relativeElev > MAX_RELATIVE_ELEVATION) continue;

      // ─ Filter: skip cells with negligible flow (dry ridges)
      if (accum < minAccum) continue;

      // ─ Filter: skip cells that ARE on the drainage channel
      if (isChannel[row][col]) continue;

      // ─ Depression detection: wider neighborhood ─
      let surroundingSum = 0;
      let surroundingCount = 0;
      let surroundingMin = Infinity;
      for (let dr = -DEPRESSION_RADIUS; dr <= DEPRESSION_RADIUS; dr++) {
        for (let dc = -DEPRESSION_RADIUS; dc <= DEPRESSION_RADIUS; dc++) {
          if (dr === 0 && dc === 0) continue;
          const nr = row + dr;
          const nc = col + dc;
          if (nr < 0 || nr >= nRows || nc < 0 || nc >= nCols) continue;
          const nElev = elevationGrid[nr][nc];
          surroundingSum += nElev;
          surroundingCount++;
          if (nElev < surroundingMin) surroundingMin = nElev;
        }
      }
      const avgSurrounding = surroundingSum / Math.max(1, surroundingCount);
      const depressionDepth = avgSurrounding - elev; // positive = cell is lower than surroundings

      // ─ Convergence: fraction of immediate 8-neighbors that are higher ─
      let higherNeighbors = 0;
      let neighborCount = 0;
      for (let d = 0; d < 8; d++) {
        const nr = row + DR[d];
        const nc = col + DC[d];
        if (nr < 0 || nr >= nRows || nc < 0 || nc >= nCols) continue;
        neighborCount++;
        if (elevationGrid[nr][nc] > elev) higherNeighbors++;
      }
      const convergence = higherNeighbors / Math.max(1, neighborCount);

      // ─ Compute individual factor scores (all normalized 0–1, higher is better) ─

      // 1. Elevation score: lower is better
      const elevationScore = 1 - relativeElev;

      // 2. Slope score: flatter is better
      const slopeScore = maxSlope > 0 ? 1 - Math.min(1, slope / maxSlope) : 1;

      // 3. Depression score: deeper depression is better
      //    Normalize by elevation range; clamp to [0, 1]
      const rawDepressionScore = depressionDepth / elevRange;
      const depressionScore = Math.max(0, Math.min(1, rawDepressionScore * 10));

      // 4. Convergence score: more neighbors higher = better (already 0–1)
      const convergenceScore = convergence;

      // 5. Catchment score: moderate-to-high accumulation (log-scaled)
      //    We want meaningful catchment but NOT the absolute maximum
      const catchmentScore = Math.log(accum + 1) / logMaxAccum;

      // 6. Channel offset score: farther from drainage channel = better
      //    Ramps linearly from 0 (on channel) to 1 (at IDEAL_CHANNEL_OFFSET)
      const channelOffsetScore = Math.min(1, distToChannel / IDEAL_CHANNEL_OFFSET);

      // ─ Composite score ─
      const score = (elevationScore     * WEIGHTS.elevation) +
                    (slopeScore          * WEIGHTS.slope) +
                    (depressionScore     * WEIGHTS.depression) +
                    (convergenceScore    * WEIGHTS.convergence) +
                    (catchmentScore      * WEIGHTS.catchment) +
                    (channelOffsetScore  * WEIGHTS.channelOffset);

      cellScores.push({
        row, col, elev, accum, slope, distToChannel,
        convergence, depressionDepth, relativeElev,
        scores: {
          elevation: round4(elevationScore),
          slope: round4(slopeScore),
          depression: round4(depressionScore),
          convergence: round4(convergenceScore),
          catchment: round4(catchmentScore),
          channelOffset: round4(channelOffsetScore),
        },
        score
      });
    }
  }

  // ── Fallback if no candidates survived the filters ──────────────────
  if (cellScores.length === 0) {
    return buildFallback(terrainModel, elevationGrid, flowAccumulation, channelDist,
                          isChannel, nRows, nCols, bounds, cellSizeLng, cellSizeLat);
  }

  // ── Step 6: Rank and select spatially-separated candidates ──────────
  cellScores.sort((a, b) => b.score - a.score);

  const minSeparation = Math.max(3, Math.floor(Math.min(nRows, nCols) * 0.08));
  const candidates = [];

  for (const cell of cellScores) {
    const tooClose = candidates.some(c => {
      const dr = Math.abs(cell.row - c.row);
      const dc = Math.abs(cell.col - c.col);
      return Math.max(dr, dc) < minSeparation;
    });

    if (!tooClose) {
      const [lng, lat] = gridToCoords(cell.row, cell.col, bounds, cellSizeLng, cellSizeLat);
      candidates.push({
        latitude: lat,
        longitude: lng,
        elevation: round2(cell.elev),
        row: cell.row,
        col: cell.col,
        suitabilityScore: round4(cell.score),
        score: round4(cell.score), // keep backward-compatible field
        flowAccumulation: Math.round(cell.accum),
        convergence: round2(cell.convergence),
        localRelief: round2(cell.depressionDepth),
        relativeElevation: round2(cell.relativeElev),
        slopeValue: round4(cell.slope),
        distanceToChannel: round2(cell.distToChannel),
        distanceToChannelMeters: round2(cell.distToChannel * cellSizeMeters),
        scoreBreakdown: cell.scores,
        reason: buildReason(cell, cellSizeMeters)
      });

      if (candidates.length >= MAX_CANDIDATES) break;
    }
  }

  // Ensure at least one candidate
  if (candidates.length === 0 && cellScores.length > 0) {
    const cell = cellScores[0];
    const [lng, lat] = gridToCoords(cell.row, cell.col, bounds, cellSizeLng, cellSizeLat);
    candidates.push({
      latitude: lat,
      longitude: lng,
      elevation: round2(cell.elev),
      row: cell.row,
      col: cell.col,
      suitabilityScore: round4(cell.score),
      score: round4(cell.score),
      flowAccumulation: Math.round(cell.accum),
      convergence: round2(cell.convergence),
      localRelief: round2(cell.depressionDepth),
      relativeElevation: round2(cell.relativeElev),
      slopeValue: round4(cell.slope),
      distanceToChannel: round2(cell.distToChannel),
      distanceToChannelMeters: round2(cell.distToChannel * cellSizeMeters),
      scoreBreakdown: cell.scores,
      reason: buildReason(cell, cellSizeMeters)
    });
  }

  return {
    selected: candidates[0],
    candidates: candidates.slice(1)
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// HELPERS
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Build a fallback selection when no candidates pass the filters.
 * Finds the best non-channel cell with highest accumulation.
 */
function buildFallback(terrainModel, elevationGrid, flowAccumulation, channelDist,
                        isChannel, nRows, nCols, bounds, cellSizeLng, cellSizeLat) {
  let bestRow = Math.floor(nRows / 2);
  let bestCol = Math.floor(nCols / 2);
  let bestScore = -Infinity;

  for (let row = 1; row < nRows - 1; row++) {
    for (let col = 1; col < nCols - 1; col++) {
      // Prefer non-channel cells with decent accumulation
      const accum = flowAccumulation[row][col];
      const onChannel = isChannel[row][col];
      const dist = channelDist[row][col];
      // Score: accumulation benefit minus channel penalty
      const s = Math.log(accum + 1) + (onChannel ? -100 : dist * 0.5);
      if (s > bestScore) {
        bestScore = s;
        bestRow = row;
        bestCol = col;
      }
    }
  }

  const [lng, lat] = gridToCoords(bestRow, bestCol, bounds, cellSizeLng, cellSizeLat);
  return {
    selected: {
      latitude: lat,
      longitude: lng,
      elevation: round2(elevationGrid[bestRow][bestCol]),
      row: bestRow,
      col: bestCol,
      suitabilityScore: 0,
      score: 0,
      flowAccumulation: Math.round(flowAccumulation[bestRow][bestCol]),
      distanceToChannel: round2(channelDist[bestRow][bestCol]),
      distanceToChannelMeters: round2(channelDist[bestRow][bestCol] * terrainModel.cellSizeMeters),
      scoreBreakdown: {},
      reason: 'Selected as best available land cell near drainage (fallback — terrain had limited valid candidates)'
    },
    candidates: []
  };
}

/**
 * Build a human-readable reason string explaining why this site was selected.
 */
function buildReason(cell, cellSizeMeters) {
  const parts = [];

  // Channel offset — the most important differentiator
  const offsetMeters = cell.distToChannel * cellSizeMeters;
  if (offsetMeters > 0) {
    parts.push(`~${Math.round(offsetMeters)}m offset from the main drainage channel (avoids stream/river)`);
  }

  // Depression
  if (cell.depressionDepth > 0) {
    parts.push(`natural depression (${cell.depressionDepth.toFixed(2)}m below surrounding terrain)`);
  }

  // Slope
  if (cell.scores.slope > 0.7) {
    parts.push('low slope (suitable for pond construction)');
  } else if (cell.scores.slope > 0.4) {
    parts.push('moderate slope');
  }

  // Convergence
  if (cell.convergence > 0.6) {
    parts.push(`terrain convergence (${Math.round(cell.convergence * 100)}% of neighbors higher)`);
  }

  // Catchment
  if (cell.accum > 10) {
    parts.push(`upstream catchment contributing area (flow accumulation: ${Math.round(cell.accum)})`);
  }

  // Elevation
  if (cell.relativeElev < 0.3) {
    parts.push('located in lower portion of terrain');
  } else if (cell.relativeElev < 0.5) {
    parts.push('moderate relative elevation');
  }

  if (parts.length === 0) {
    parts.push('best composite terrain suitability score for land-based pond construction');
  }

  return 'Land site selected: ' + parts.join('; ') + `.`;
}

function round2(v) { return Math.round(v * 100) / 100; }
function round4(v) { return Math.round(v * 10000) / 10000; }

module.exports = { selectPondSite, WEIGHTS };

File: python_server/services/pond_site_selection.py
"""
Pond Site Selection Service

Identifies the best LAND location for a village pond using multi-factor
terrain suitability scoring. The selected site is adjacent to the drainage
system, NOT on the main channel/river itself.

Scoring factors (weights sum to 1.0):
  - channelOffset  0.25  Penalise cells on or near drainage channels
  - depression     0.20  Prefer natural terrain bowls
  - catchment      0.20  Prefer meaningful upstream flow
  - slope          0.15  Prefer flat / low-gradient areas
  - elevation      0.10  Prefer lower relative terrain
  - convergence    0.10  Prefer areas surrounded by higher ground

Identical algorithm to pondSiteSelection.js — only Python syntax differs.
"""

import math
import numpy as np
from collections import deque
from .terrain_analysis import DR, DC, grid_to_coords

# ── Configurable weights ──────────────────────────────────────────────────────
WEIGHTS = {
    'elevation':     0.10,
    'slope':         0.15,
    'depression':    0.20,
    'convergence':   0.10,
    'catchment':     0.20,
    'channelOffset': 0.25,
}

CHANNEL_PERCENTILE   = 0.02   # top 2 % of accum = drainage channel
MIN_ACCUM_FRACTION   = 0.005  # 0.5 % of max accum required
MAX_RELATIVE_ELEV    = 0.65   # skip upper 35 % of terrain
DEPRESSION_RADIUS    = 3      # neighbourhood for depression detection
IDEAL_CHANNEL_OFFSET = 5      # cells this far from channel → full offset score
MAX_CANDIDATES       = 5


def select_pond_site(terrain_model: dict, metadata: dict) -> dict:
    """
    Score every non-channel grid cell and return the best pond site + candidates.
    Returns: { 'selected': {...}, 'candidates': [...] }
    """
    elev_grid = terrain_model['elevation_grid']
    flow_accum = terrain_model['flow_accumulation']
    n_rows     = terrain_model['nRows']
    n_cols     = terrain_model['nCols']
    bounds     = terrain_model['bounds']
    cell_size_lng = terrain_model['cellSizeLng']
    cell_size_lat = terrain_model['cellSizeLat']
    cell_size_m   = terrain_model['cellSizeMeters']

    elev_min   = metadata['minElevation']
    elev_max   = metadata['maxElevation']
    elev_range = (elev_max - elev_min) or 1.0

    # ── 1. Channel mask ───────────────────────────────────────────────────
    accum_flat = flow_accum.ravel()
    sorted_accum = np.sort(accum_flat)
    threshold_idx = int(len(sorted_accum) * (1 - CHANNEL_PERCENTILE))
    channel_threshold = sorted_accum[min(threshold_idx, len(sorted_accum) - 1)]
    max_accum = float(sorted_accum[-1])

    is_channel = flow_accum >= channel_threshold
    print(f"  Accum max={max_accum:.0f}, channel threshold={channel_threshold:.0f}")

    # ── 2. BFS distance-to-nearest-channel ───────────────────────────────
    channel_dist = _bfs_channel_distance(is_channel, n_rows, n_cols)

    # ── 3. Slope grid ─────────────────────────────────────────────────────
    slope_grid, max_slope = _compute_slope(elev_grid, n_rows, n_cols)

    # ── 4. Score every candidate cell ─────────────────────────────────────
    log_max_accum = math.log(max_accum + 1)
    min_accum     = max_accum * MIN_ACCUM_FRACTION
    margin        = max(DEPRESSION_RADIUS + 1, int(min(n_rows, n_cols) * 0.05))

    cell_scores = []
    for row in range(margin, n_rows - margin):
        for col in range(margin, n_cols - margin):
            elev  = float(elev_grid[row, col])
            accum = float(flow_accum[row, col])

            rel_elev = (elev - elev_min) / elev_range
            if rel_elev > MAX_RELATIVE_ELEV:
                continue
            if accum < min_accum:
                continue
            if is_channel[row, col]:
                continue

            # Depression (wider neighbourhood)
            surr = _neighbourhood_stats(elev_grid, row, col, DEPRESSION_RADIUS, n_rows, n_cols)
            depression_depth = surr['mean'] - elev  # positive → cell is below surroundings

            # Convergence (immediate 8 neighbours)
            higher = sum(
                1 for d in range(8)
                if 0 <= row + int(DR[d]) < n_rows and 0 <= col + int(DC[d]) < n_cols
                and float(elev_grid[row + int(DR[d]), col + int(DC[d])]) > elev
            )
            convergence = higher / 8.0

            slope = float(slope_grid[row, col])
            dist_to_ch = float(channel_dist[row, col])

            # Individual scores (0–1, higher = better)
            s_elev     = 1.0 - rel_elev
            s_slope    = (1.0 - min(1.0, slope / max_slope)) if max_slope > 0 else 1.0
            s_depr     = min(1.0, max(0.0, (depression_depth / elev_range) * 10))
            s_conv     = convergence
            s_catch    = math.log(accum + 1) / log_max_accum
            s_offset   = min(1.0, dist_to_ch / IDEAL_CHANNEL_OFFSET)

            score = (s_elev   * WEIGHTS['elevation']     +
                     s_slope  * WEIGHTS['slope']         +
                     s_depr   * WEIGHTS['depression']    +
                     s_conv   * WEIGHTS['convergence']   +
                     s_catch  * WEIGHTS['catchment']     +
                     s_offset * WEIGHTS['channelOffset'])

            cell_scores.append({
                'row': row, 'col': col,
                'elev': elev, 'accum': accum,
                'slope': slope, 'dist_to_ch': dist_to_ch,
                'convergence': convergence,
                'depression_depth': depression_depth,
                'rel_elev': rel_elev,
                'score': score,
                'scores': {
                    'elevation':     _r4(s_elev),
                    'slope':         _r4(s_slope),
                    'depression':    _r4(s_depr),
                    'convergence':   _r4(s_conv),
                    'catchment':     _r4(s_catch),
                    'channelOffset': _r4(s_offset),
                }
            })

    # ── 5. Fallback ───────────────────────────────────────────────────────
    if not cell_scores:
        return _fallback(terrain_model, is_channel, channel_dist, n_rows, n_cols,
                         bounds, cell_size_lng, cell_size_lat, cell_size_m)

    # ── 6. Rank + spatially separate candidates ───────────────────────────
    cell_scores.sort(key=lambda x: x['score'], reverse=True)
    min_sep = max(3, int(min(n_rows, n_cols) * 0.08))

    candidates = []
    for cell in cell_scores:
        too_close = any(
            max(abs(cell['row'] - c['row']), abs(cell['col'] - c['col'])) < min_sep
            for c in candidates
        )
        if not too_close:
            candidates.append(cell)
        if len(candidates) >= MAX_CANDIDATES:
            break

    if not candidates:
        candidates = [cell_scores[0]]

    def build_site(cell):
        lng, lat = grid_to_coords(cell['row'], cell['col'], bounds, cell_size_lng, cell_size_lat)
        return {
            'latitude':               round(lat, 8),
            'longitude':              round(lng, 8),
            'elevation':              _r2(cell['elev']),
            'row':                    cell['row'],
            'col':                    cell['col'],
            'suitabilityScore':       _r4(cell['score']),
            'score':                  _r4(cell['score']),
            'flowAccumulation':       round(cell['accum']),
            'convergence':            _r2(cell['convergence']),
            'localRelief':            _r2(cell['depression_depth']),
            'relativeElevation':      _r2(cell['rel_elev']),
            'distanceToChannel':      _r2(cell['dist_to_ch']),
            'distanceToChannelMeters': _r2(cell['dist_to_ch'] * cell_size_m),
            'scoreBreakdown':         cell['scores'],
            'reason':                 _build_reason(cell, cell_size_m),
        }

    selected   = build_site(candidates[0])
    alternates = [build_site(c) for c in candidates[1:]]
    return {'selected': selected, 'candidates': alternates}


# ─── Helpers ─────────────────────────────────────────────────────────────────

def _bfs_channel_distance(is_channel: np.ndarray, n_rows: int, n_cols: int) -> np.ndarray:
    dist = np.full((n_rows, n_cols), np.inf, dtype=np.float32)
    q = deque()
    for r in range(n_rows):
        for c in range(n_cols):
            if is_channel[r, c]:
                dist[r, c] = 0.0
                q.append((r, c))

    while q:
        r, c = q.popleft()
        for d in range(8):
            nr, nc = r + int(DR[d]), c + int(DC[d])
            if 0 <= nr < n_rows and 0 <= nc < n_cols:
                step = 1.414 if d % 2 == 1 else 1.0
                nd = dist[r, c] + step
                if nd < dist[nr, nc]:
                    dist[nr, nc] = nd
                    q.append((nr, nc))
    return dist


def _compute_slope(elev: np.ndarray, n_rows: int, n_cols: int):
    slope = np.zeros((n_rows, n_cols), dtype=np.float64)
    for d in range(8):
        dr, dc = int(DR[d]), int(DC[d])
        dist = 1.414 if d % 2 == 1 else 1.0
        for row in range(n_rows):
            for col in range(n_cols):
                nr, nc = row + dr, col + dc
                if 0 <= nr < n_rows and 0 <= nc < n_cols:
                    g = abs(float(elev[row, col]) - float(elev[nr, nc])) / dist
                    if g > slope[row, col]:
                        slope[row, col] = g
    max_slope = float(slope.max())
    return slope, max_slope


def _neighbourhood_stats(elev: np.ndarray, row: int, col: int, radius: int,
                          n_rows: int, n_cols: int) -> dict:
    vals = []
    for dr in range(-radius, radius + 1):
        for dc in range(-radius, radius + 1):
            if dr == 0 and dc == 0:
                continue
            nr, nc = row + dr, col + dc
            if 0 <= nr < n_rows and 0 <= nc < n_cols:
                vals.append(float(elev[nr, nc]))
    mean = sum(vals) / len(vals) if vals else float(elev[row, col])
    return {'mean': mean}


def _build_reason(cell: dict, cell_size_m: float) -> str:
    parts = []
    offset_m = cell['dist_to_ch'] * cell_size_m
    if offset_m > 0:
        parts.append(f"~{round(offset_m)}m offset from the main drainage channel (avoids stream/river)")
    if cell['depression_depth'] > 0:
        parts.append(f"natural depression ({cell['depression_depth']:.2f}m below surrounding terrain)")
    if cell['scores']['slope'] > 0.7:
        parts.append("low slope (suitable for pond construction)")
    elif cell['scores']['slope'] > 0.4:
        parts.append("moderate slope")
    if cell['convergence'] > 0.6:
        parts.append(f"terrain convergence ({round(cell['convergence'] * 100)}% of neighbors higher)")
    if cell['accum'] > 10:
        parts.append(f"upstream catchment contributing area (flow accumulation: {round(cell['accum'])})")
    if cell['rel_elev'] < 0.3:
        parts.append("located in lower portion of terrain")
    elif cell['rel_elev'] < 0.5:
        parts.append("moderate relative elevation")
    if not parts:
        parts.append("best composite terrain suitability score for land-based pond construction")
    return "Land site selected: " + "; ".join(parts) + "."


def _fallback(terrain_model, is_channel, channel_dist, n_rows, n_cols,
               bounds, cell_size_lng, cell_size_lat, cell_size_m):
    elev_grid  = terrain_model['elevation_grid']
    flow_accum = terrain_model['flow_accumulation']
    best_r, best_c, best_s = n_rows // 2, n_cols // 2, -math.inf
    for r in range(1, n_rows - 1):
        for c in range(1, n_cols - 1):
            s = math.log(float(flow_accum[r, c]) + 1) + (-100 if is_channel[r, c] else float(channel_dist[r, c]) * 0.5)
            if s > best_s:
                best_s, best_r, best_c = s, r, c
    lng, lat = grid_to_coords(best_r, best_c, bounds, cell_size_lng, cell_size_lat)
    return {
        'selected': {
            'latitude': round(lat, 8), 'longitude': round(lng, 8),
            'elevation': _r2(float(elev_grid[best_r, best_c])),
            'row': best_r, 'col': best_c,
            'suitabilityScore': 0, 'score': 0,
            'flowAccumulation': round(float(flow_accum[best_r, best_c])),
            'distanceToChannelMeters': _r2(float(channel_dist[best_r, best_c]) * cell_size_m),
            'scoreBreakdown': {},
            'reason': 'Selected as best available land cell near drainage (fallback — limited valid candidates)',
        },
        'candidates': []
    }


def _r2(v): return round(v * 100) / 100
def _r4(v): return round(v * 10000) / 10000
